samcli/lib/runtimes/base.py

```diff
@@ -0,0 +1,342 @@
+
+
+@unique
+class Family(FamilyDataMixin, Enum):
+    DOTNET = (
+        "dotnet",
+        "dotnet",
+        [("cli-package", _templates / "cookiecutter-aws-sam-hello-dotnet")],
+        True,
+    )
+    GO = "go", None, [("mod", _templates / "cookiecutter-aws-sam-hello-golang")], False, "Go1"
+    JAVA = (
+        "java",
+        "java",
+        [
+            ("maven", _templates / "cookiecutter-aws-sam-hello-java-maven"),
+            ("gradle", _templates / "cookiecutter-aws-sam-hello-java-gradle"),
+        ],
+        True,
+        "Java8",
+    )
+    NODEJS = (
+        "nodejs",
+        "nodejs",
+        [("npm", _templates / "cookiecutter-aws-sam-hello-nodejs")],
+        True,
+    )
+    PROVIDED = "provided", "", [], False, None, False  # type: ignore [var-annotated]
+    PYTHON = (
+        "python",
+        "python",
+        [("pip", _templates / "cookiecutter-aws-sam-hello-python")],
+        True,
+        "Python36",
+    )
+    RUBY = "ruby", "ruby/lib", [("bundler", _templates / "cookiecutter-aws-sam-hello-ruby")]
+
+
+@dataclass
+class RuntimeDataMixin:
+    """
+    Definition of a SAM Runtime
+    """
+
+    key: str
+    family: Family
+    lambda_image: Optional[str] = None
+    archs: List[Architecture] = field(default_factory=list)
+    is_lambda_enum: bool = True
+    # build_workflow: WorkflowConfig
+    # workflow_selector: ???
+
+    # debug settings - samcli/local/docker/lambda_debug_settings.py
+    # debug_entry_setting: callable
+
+    # runtime enum - samcli/local/docker/lambda_image.py
+
+    # DEPRECATED_RUNTIMES - samcli/lib/build/constants.py
+    is_deprecated: bool = False
+
+    # local manifest - samcli/lib/init/local_manifest.json
+    # is local manifest useful at all?
+    def archs_as_list_of_str(self) -> List[str]:
+        return [a.value for a in self.archs]
+
+    @property
+    def is_provided(self):
+        return self.family == Family.PROVIDED
+
+    def to_local_init_manifest(self) -> List[Dict]:
+        """
+        Example:
+        "java8.al2": [
+            {
+                "directory": "template/cookiecutter-aws-sam-hello-java-gradle",
+                "displayName": "Hello World Example: Gradle",
+                "dependencyManager": "gradle",
+                "appTemplate": "hello-world",
+                "packageType": "Zip",
+                "useCaseName": "Hello World Example"
+            },
+            {
+                "directory": "template/cookiecutter-aws-sam-hello-java-maven",
+                "displayName": "Hello World Example: Maven",
+                "dependencyManager": "maven",
+                "appTemplate": "hello-world",
+                "packageType": "Zip",
+                "useCaseName": "Hello World Example"
+            }
+        ]
+        """
+        raise NotImplementedError()
+
+
+class RuntimeEnumBase(RuntimeDataMixin, Enum):
+    @classmethod
+    def from_str(cls, runtime: Optional[str]) -> "RuntimeEnumBase":
```

Comment:
The return type has to be in quotes because python doesn't like you returning an instance from a class method?

Reply:
What I wanted to do is to return the self type.
The type checker didn't like it somehow

Seems like there's a `Self` type but it's available in 3.11+ https://peps.python.org/pep-0673/

Reply:
Looks like you might be able to do something using generic bounds like
```
Self = TypeVar("Self", bound="Shape")

class Shape:
    @classmethod
    def from_config(
        cls: type[Self], config: dict[str, float]
    ) -> Self:
        return cls(config["scale"])
```

though I don't particularly think that's any more intuitive than what you have now